### DISCUSSION TITLE
Add 24h format for {time-locale}

### DISCUSSION THREAD

**Discussion by iG8R** [gildas-lormeau/SingleFile#1302] [General]
Please, add 24h format for {time-locale}.

**Comment by Codereamp** [comment]
I'd also append a suggestion to 

- allow different {month, day, years} orders
- allow some padding to time and date parts (so april 4 can become 04)

If there are more than one files saved this features might improve sorting issues

I'm not sure about posting this as a separate suggestion, maybe the author would think about a little more complex system with formatting to address maybe also some other suggestions :)

**Reply by gildas-lormeau** [reply]
You can do all of that with the existing template functions if I am not wrong, e.g. `%pad-left<{month-locale}|1| >`.

**Reply by Codereamp** [reply]
@gildas-lormeau , thanks for pointing out. The help page is a big help as I see, but it is somehow distanced from the user at least on my firefox 126.0b5
- When it shown, Ctrl-F doesn't search at all
- The top local links like "Template functions" dont't work. 
- You insert links like "this list" into infoboxes, but they open the help file and don't jump into the corresponding section. 

This all goes away if ones just "Right click-save as" the help page and use it separately, but if for some reason those issues are easily fixed, it will help a lot. Thanks! 

Additional note. For some reason the help file is not publicly available to be reachable by googling. I know that it's the part of the sources on github, but as I noticed, many items from github is not reachable. This works like a paradox of the modern times. If user wants advanced topics he or she rather google [SingleFile help] than look at the option to find a "help" link. If it's possible somehow, please make the help file reachable by google. As a test, on exact search ["The template functions are used to format"] google says "nothing found"

**Reply by gildas-lormeau** [reply]
@Codereamp The issues you're describing are strange. I cannot reproduce them in Firefox on macOS. Ctrl-F and links work as expected in the options page and in the help page.

I agree that it would be helpful to find the help page more easily on the web. I plan do that when I will use a custom domain to promote the extension.

**Comment by Codereamp** [comment]
> @Codereamp The issues you're describing are strange. I cannot reproduce them in Firefox on macOS. Ctrl-F and links work as expected in the options page and in the help page.

Thanks for checking. Based on your observation I suspected this has something to do with my changed settings. It is actually so. I have a config change when `browser.link.open_newwindow` setting always set to 1, so I when I need a new tab/window I chose this explicitly regardless of the target="_blank" attribute. Since your options page is a general page for the browser (apart from minor differences), and you probably use this attribute to separate the entities, I see this strange "beast" - the help page opened not in a new tab, but inside the option pane and this one behaves not pleasantly as I described. The screenshot is below. Probably as the extension developer you can't change this so somehow make a super-override that ignores this setting. Maybe it's a side effect/issue on the Mozilla side that should be addressed somehow


![SingleFileHelp](https://github.com/gildas-lormeau/SingleFile/assets/41295076/a0614d8a-ccc4-4634-96f5-bcfb97eb1b9a)

**Comment by Codereamp** [comment]
> Please, add 24h format for {time-locale}.

As for the original question, I did a little digging in the code and at the web. It looks like `toLocaleDateString` `toLocaleTimeString` `toLocaleString` used for several template variables without parameters so they rely on what the page provided and the preference of the user combined. It means that no matter what, user can not force his or her preferred format in this case. 

My possible suggestions are 
- to use `window.navigator.language` as a parameter in those three calls. Might break existing behavior for current users. 
- introduce a three new variables that uses the browser `window.navigator.language`
- introduce the date/time locale variable in the parameters, empty by default so if not changed, keeps the calls as they are now (ideal for compatibility). But if it is changed, the three function would accept it to format the date/time. (UPDATE: this is my personal preference amongst the three, because as a software developer I prefer English almost everywhere, but as a non-native english speaker I prefer something else for date/time :) )

**Comment by gildas-lormeau** [comment]
Regarding the help page, I don't think I will be able to find a way to circumvent the navigation issues when `browser.link.open_newwindow` is set to `true`. Maybe, you could create a ticket on https://bugzilla.mozilla.org.

I've added the functions `%date-locale`, `%time-locale`, `%datetime-locale` which require as parameter a language string (e.g. `%time-locale<en-US>`) and the template variable `{navigator-language}`. They will available in the next version.

**Comment by Codereamp** [comment]
checked new locale additions both in Firefox and Chrome. Thanks, @gildas-lormeau, for them and for the add-on itself. For me it was critical to find a good "year first, month next, ...24h" country-locale, so the Sweden worked like a charm, both for Swedish and English. en-SE also [mentioned](https://www.reddit.com/r/ISO8601/comments/khdyrn/to_get_the_closest_to_iso_8601_on_your_device/?rdt=63634) as the closest to ISO when used in OSes settings 

So, the first one
`%if-empty<{page-title}|No title>(%datetime-locale<sv-SE>).html`
 .... (2024-05-21 20_33_10)...

and the second differs only with a comma
`%if-empty<{page-title}|No title>(%datetime-locale<en-SE>).html`
 ....(2024-05-21, 20_33_10)...

**Reply by gildas-lormeau** [reply]
Thanks you the info and the feedback!
